rfc/graph-feature-flags.md

```diff
@@ -0,0 +1,307 @@
+```
+
+Let's say, we'd like to introduce a new feature in the staging environment that will change the behavior of the API dramatically.
+We don't want to affect other developers who might be integration testing other features in the staging environment.
+As such, we want to publish a new Subgraph Schema under a feature flag.
+
+We can add a `--feature-flag` flag to the `subgraph publish` command:
+
+```bash
+npx wgc subgraph publish products --schema ../demo/subgraphs/products/products.graphql --routing-url http://localhost:4002/graphql --label=team=A --feature-flag=experimental --disable-composition-checks
+```
+
+We've deployed a copy of the API to a different port, and we've published a new Subgraph Schema under the feature flag `products.experimental`.
+We've also disabled composition checks for this Subgraph, as we're testing a new feature that might not be compatible with changes that other developers are making.
+
+To be able to test this new feature end-to-end, we need to enable the feature flag in our client.
+We can achieve this by setting a cookie:
+
+```bash
+curl -X POST http://localhost:4002/graphql -H "Content-Type: application/json" -H "Cookie: X-WG-Feature-Flag=products.experimental" --data '{ "query": "{ products { id name } }" }'
+```
+
+This will send a request to the staging environment with the feature flag `products.experimental` enabled.
+The Router in the staging environment will pick up the feature flag and use the "Graph Resolver" with the feature enabled instead of the default one.
+
+Once the feature has been tested and the team has confidence in it,
+the Subgraph Schema can be published without the feature flag,
+and we can remove the feature flag from the staging environment:
+
+```bash
+npx wgc subgraph publish products --schema ../demo/subgraphs/products/products.graphql --routing-url http://localhost:4001/graphql --label=team=A --remove-feature-flags
+```
+
+Now this feature is available to all users of the staging environment.
+
+## Alternative ways to expose & enable Feature Flags
+
+It would also be possible to make all available feature flag introspectable using the `__features` field in the `Query` type.
+This would allow clients, like Cosmo Studio, to show all available feature flags with a description and a toggle to enable or disable them.
+
+```graphql
+type Query {
+  __features: [FeatureFlag!]!
+}
+
+type FeatureFlag {
+  name: String!
+  subgraph: String!
+  description: String!
+}
+```
+
+Query:
+
+```graphql
+{
+  __features {
+    name
+    subgraph
+    description
+  }
+}
+```
+
+Example response:
+
+```json
+{
+  "__features": [
+    {
+      "name": "experimental",
+      "subgraph": "products",
+      "description": "Enables the experimental feature for the products Subgraph"
+    }
+  ]
+}
+```
+
+## Set feature flags using the extensions field in the GraphQL Request
+
+It would also be possible to set feature flags using the `extensions` field in the GraphQL Request.
```

Comment:
How will this apply to responses, specifically with errors?

Reply:
Can you elaborate on this?

Reply:
I know there's the other RFC around responses including subgraph errors with extensions, I was wondering if the feature flag being used would also be returned for an error to know the error was from the subgraph with the flag enabled.